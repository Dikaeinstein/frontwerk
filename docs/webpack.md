# frontwerk 🛠️ builds with Webpack

_Details coming soon_

[Read the source](https://github.com/tricinel/frontwerk/blob/master/src/scripts/webpack.js)
