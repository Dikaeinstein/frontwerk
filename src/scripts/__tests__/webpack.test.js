/* eslint-disable global-require */
import {
  unquoteSerializer,
  winPathSerializer
} from '../../helpers/serializers';

jest.mock('cross-spawn');
jest.mock('concurrently');
jest.mock('jest');

const cases = require('jest-in-case');
const jestSerializerPath = require('jest-serializer-path');

expect.addSnapshotSerializer(unquoteSerializer);
expect.addSnapshotSerializer(winPathSerializer);
expect.addSnapshotSerializer(jestSerializerPath);

const testCases = [
  {
    name: 'calls webpack CLI with default args'
  },
  {
    name: '--no-clean will not clean the dist directory',
    args: ['--no-clean']
  },
  {
    name: '--env will pass the environment',
    args: ['--env', 'production']
  },
  {
    name: '--mode will pass the development mode',
    args: ['--mode', 'development']
  },
  {
    name: '--mode will pass the production mode',
    args: ['--mode', 'production']
  },
  {
    name: '--entry will pass the entry folder',
    args: ['--entry', 'src']
  },
  {
    name: '--output-path will pass the output path folder',
    args: ['--output-path', './.tmp/build']
  },
  {
    name: 'does not use built-in config with webpack.config.js file',
    fileExists: filename => filename === 'webpack.config.js'
  },
  {
    name: 'does not use built-in config with --config',
    args: ['--config', './custom-config.js']
  }
];

const testFn = ({ fileExists = () => false, args = [] }) => {
  const { sync: crossSpawnSyncMock } = require('cross-spawn');
  const originalExit = process.exit;
  const originalArgv = process.argv;
  process.exit = jest.fn();

  Object.assign(require('../../utils/fileExists'), { fileExists });
  Object.assign(require('../../utils/resolveBin'), {
    resolveBin: (modName, { executable = modName } = {}) => executable
  });

  process.exit = jest.fn();

  try {
    // tests
    process.argv = ['node', '../webpack', ...args];
    crossSpawnSyncMock.mockClear();

    require('../webpack');

    expect(crossSpawnSyncMock).toHaveBeenCalledTimes(1);
    const [firstCall] = crossSpawnSyncMock.mock.calls;
    const [script, calledArgs] = firstCall;
    expect([script, ...calledArgs].join(' ')).toMatchSnapshot();
  } catch (error) {
    throw error;
  } finally {
    // afterEach
    process.exit = originalExit;
    process.argv = originalArgv;
    jest.resetModules();
  }
};

cases('webpack', testFn, testCases);
