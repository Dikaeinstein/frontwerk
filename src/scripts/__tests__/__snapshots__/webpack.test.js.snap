// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format --entry will pass the entry folder 1`] = `"concurrently --names webpack \\"cross-env BUILD_WEBPACK=true BUILD_ENTRY=src webpack --config ./src/config/webpack.config.js\\""`;

exports[`format --env will pass the environment 1`] = `"concurrently --names webpack \\"cross-env BUILD_WEBPACK=true --env production webpack --config ./src/config/webpack.config.js\\""`;

exports[`format --no-clean will not clean the dist directory 1`] = `"concurrently --names webpack \\"cross-env BUILD_WEBPACK=true webpack --config ./src/config/webpack.config.js\\""`;

exports[`format --output-path will pass the output path folder 1`] = `"concurrently --names webpack \\"cross-env BUILD_WEBPACK=true BUILD_OUTPUT_PATH=./.tmp/build webpack --config ./src/config/webpack.config.js\\""`;

exports[`format calls webpack CLI with default args 1`] = `"concurrently --names webpack \\"cross-env BUILD_WEBPACK=true webpack --config ./src/config/webpack.config.js\\""`;

exports[`format does not use built-in config with --config 1`] = `"concurrently --names webpack \\"cross-env BUILD_WEBPACK=true webpack\\""`;

exports[`format does not use built-in config with webpack.config.js file 1`] = `"concurrently --names webpack \\"cross-env BUILD_WEBPACK=true webpack\\""`;

exports[`format watches with --watch 1`] = `"concurrently --names webpack \\"cross-env BUILD_WEBPACK=true webpack --config ./src/config/webpack.config.js --watch\\""`;
